man/man3/scsi.3
.TH SCSI 3
.SH NAME
openscsi, scsiready, scsi, scsicmd, scsierror \- SCSI device operations
.SH SYNOPSIS
.nf
.ft L
#include <u.h>
#include <libc.h>
#include <disk.h>
.ft
.PP
.ft L
typedef struct Scsi {
	char  *inquire;
	int   rawfd;
	int   nchange;
	ulong changetime;
};
.ft
.PP
.B
Scsi* openscsi(char *devdir)
.PP
.B 
void  closescsi(Scsi *s)
.PP
.B
int   scsiready(Scsi *s)
.PP
.B
int   scsi(Scsi *s, uchar *cmd, int ncmd,
.br
                    void *data, int ndata, int dir)
.PP
.B
int   scsicmd(Scsi *s, uchar *cmd, int ncmd,
.br
                    void *data, int ndata, int dir)
.PP
.B
char* scsierror(int asc, int ascq)
.PP
.B 
int   scsiverbose;
.SH DESCRIPTION
These routines provide an interface
to a SCSI or ATAPI device via Plan 9's
\fIsd\fR(3).
.PP
.I Openscsi
attempts to open the file
.IB devdir /raw
and use it to send raw SCSI commands.
On success, it reads the device's inquiry
string and stores it in in returned
.B Scsi
structure.
.I Closescsi
closes the connection and frees the
.B Scsi
structure.
.PP
.I Scsiready
sends the ``unit ready'' command up to three times,
returning zero if the unit responds that it is ready,
or \-1 on error.
.PP
.I Scsierror
returns a textual description of the SCSI status
denoted by the ASC and ASCQ sense codes.
The description is found by consulting
.BR /sys/lib/scsicodes .
The returned string will be overwritten by
the next call to
.IR scsierror .
.PP
.I Scsi
and
.I scsicmd
execute a single SCSI command on the named device.
There should be 
.I ncmd
bytes of
command data in 
.IR cmd ;
if
.I dir
is 
.BR Sread ,
a successful operation
will store up to
.I ndata
bytes into
.IR data ,
returning the number of bytes stored.
If
.I dir
is
.BR Swrite ,
the 
.I ndata
bytes beginning at
.I data
are transmitted as the data argument to
the command, and the
number of bytes written is returned.
If
.I dir
is
.BR Snone ,
.I data
and
.I ndata
are ignored.
On error, 
.I scsi
and
.I scsicmd
return \-1.
.I Scsicmd
simply issues the command and
returns the result;
.I scsi
works a bit harder and
is the more commonly used routine.
.I Scsi
attempts to send the command;
if it is successful, 
.I scsi
returns what
.I scsicmd
returned.
Otherwise, 
.I scsi
sends a request sense command to
obtain the reason for the failure,
sends a unit ready command in
an attempt to bring the unit out of any
inconsistent states, and tries again.
If the second try fails,
.I scsi 
sends the request
sense and unit ready commands
again
and then uses
.I scsierror
to set
.I errstr
with a reason for failure.
.PP
The 
.B nchange
and
.B changetime
fields
in the
.B Scsi
structure
record the number of times a media
change has been detected, and the
time when the current media was
inserted into the drive (really the
first time a SCSI command was issued
after it was inserted).
They are maintained by 
.IR scsi .
.PP
If
.I scsiverbose
is set,
these commands will produce a fair
amount of debugging output on file descriptor 2
when SCSI commands fail.
.SH FILES
.TP
.B /sys/lib/scsicodes
List of textual messages corresponding to SCSI error codes;
consulted by
.BR scsierror .
.SH SOURCE
.B \*9/src/libdisk/scsi.c
.SH SEE ALSO
Plan 9's
\fIsd\fR(3) and
\fIscuzz\fR(8)
.SH BUGS
SCSI devices on Unix do not present the interface expected by
these routines.


CHANGES
November 26, 2005
	9term: add button 3 menu
	include: move utf.h, fmt.h from u.h to libc.h
	dist: print summary w/ errors during INSTALl

November 25, 2005
	root: add Makefile and configure for the blind,
		rename NOTES to README	

November 21, 2005
	lib9, 9pserve, vac: 9P2000.u fixes (Lucho Ionkov)

November 15, 2005
	libdraw: more X11 fixes
	libc.h: sched_yield NetBSD tweak (Anselm Garbe)

November 12, 2005
	faces: fix face location
	jpg, strings: remove references to /fd/

November 12, 2005
	various: X11 fixes

November 11, 2005
	9term: more fixes (Tim Wiess)

November 10, 2005
	9term: handle \r, other fixes (Tim Wiess)
	auxstats: new Darwin tweaks

November 9, 2005
	libthread: fix OS X context switch code

November 7, 2005
	vbackup: various fixes

November 6, 2005
	9term: new port directly from rio
	include/libc.h: p9syslog, not syslog in varargck
	libauthsrv: close fd, memory leak in authdial
	libdraw: X11 debugging helps
	lib9: add getcallerpc-arm (Ron Minnich)
	auxstats: Fix FreeBSD 6 build (William Josephson)

November 4, 2005
	stats: add -C to argchars (Lou Kamenov)
	libthread: ARM fixes (Ron Minnich), getcontext fixes

November 2, 2005
	vbackup: longer timeout on BSD mount
	libthread: Linux ARM fixes (Ron Minnich)

November 1, 2005
	various: ctype sign fixes (Tom Miller)
	libfmt: avoid overflow on %f
	libthread: Linux ARM fixes (Ron Minnich), FreeBSD 5 fixes
	libventi: various cleanup

October 31, 2005
	acme: preparation for multiline tags (Paul Lalonde)
	bc: 64-bit safe (Taj Khattra)
	faces: initial port (John Cummings)
	stats: add -F option (Lou Kamenov)
	various: ctype sign fixes (Tom Miller)

October 30, 2005
	auxstats: more OS X fixes (Jeff Sickel)	

October 29, 2005
	9p: clunk fids before exiting
	mk: handle MKSHELL type correctly
	upas, acme mail: add initial port (John Cummings)
	vbackup: silence compiler warnings
	venti: silence compiler warnings
	lib9: remove reference to ffork
		print correct file name in truerand error
	libdisk: add Disk routines
	libdiskfs: silence compiler warnings
	libframe: add noredraw flag
	libhtml: many fixes
	libsunrpc: silence compiler warnings

October 18, 2005
	9l: add autoframework (like autolib) for OS X (Jeff Sickel)
	page: use "9 file" always
	ps, psu: narrower display
	stats: better OS X support (Jeff Sickel)

October 17, 2005
	cb: undef old type macros before redefining.

October 5, 2005
	stats: better OS X support (Jeff Sickel) 

September 30, 2005
	libhtml: more &names and numbers (Erik Quanstrom)

September 29, 2005
	libc.h: add #pragmas for varargck, #ifdef'ed out.

September 26, 2005
	libthread: print thread info on SIGQUIT, SIGINFO, logging
	acme: add more thread names

September 21, 2005
	9term: fix SunOS typo (Bengt Kleberg, glaive)

September 19, 2005
	libhtml: more &names (Erik Quanstrom)

September 19, 2005
	lib9/truerand: OpenBSD lets you open /dev/random
		but not read it! (Chad Dougherty, Mechiel Lukkien)

September 15, 2005
	libdraw/x11-event: Ignore unexpected events (Rob Pike)
	jpg: do not revert to CMAP for second file (Rob Pike)

September 13, 2005
	9pserve: signedness bug

September 12, 2005
	dist: various attempts at Debian packaging
	lib/moveplan9.sh: allow specification of old directory
	Getdir: interpret tags with $ in them
	mpm: use g++ as LD, add tmac.pm to tmac directory
	include/libc.h, fcall.h: add 9P2000.u extensions
	9pserve: add 9P2000.u extensions
	vac: add 9P2000.u extensions, signedness bug

September 9, 2005
	lc: add -p like man page says
	various: insert (uchar) casts for ctype arguments
		for broken NetBSD.

September 4, 2005
	dist/main.html: note x86 OpenBSD

August 31, 2005
	file: spell IA-64 correctly (Zoltan Jarai)

August 30, 2005
	dist/debian, dist/pkg: start playing with building packages
	dist/isum: ignore 'loop not entered at top' (SunOS cc)
	dist/ss.html: add screen shots
	dist/unix.html: update VN license summary
	9l: no need to handle OpenBSD specially
	file: handle #! scripts, ELF better
	font/lucidasans/bold*unicode.7.font: more subfonts
	g: search .pl, .cgi
	hget: add hget(1)
	lp: add preliminary lp (not expected to work yet)
	src: use Plan 9 file(1)

August 12, 2005
	lib/moveplan9.files: add bin/stack
	include/libc.h: use DM constants from v9fs, add DMTMP
	libthread: add Linux-power-asm.s
	cb: add

August 11, 2005
	various: silence warnings
	unix/*: note new BSD license from Vita Nuova
	libdiskfs: add UFS2 support
	libthread: add NetBSD power assembly
	9term: send tty's current interrupt character on DEL keystroke
	rc: add limited rfork command, make sure errors propagate
	venti(8): document bloom line in config
	vftp: add get command
	vnfs: use strtoull for arguments instead of atoi.

August 10, 2005
	src/cmd/mkfile: do not build lp yet
	dist/addsrclinks, fixurls: use /usr/bin/env perl to find perl (NetBSD)
	lib9/_p9dir: avoid 32-bit overflow in old Linux size calculation
	lib9/truerand: use /dev/srandom on OpenBSD
	libthread: annoying wrappers for OpenBSD
	libmp: different assembly comments for old GNU as
	9p: set exit status
	file: add, appease gcc
	getflags: added
	mk: decide mtime-related race conservatively
	postscript/text2post: added
	rcmain: make fn cd safer for use with rc -e (in mk)
	rio: turn off debugging prints, remove 9wm.man, fixes from Andrey
	vbackup: NetBSD fixes for mount-BSD.c
	venti/srv: add <sys/stat.h> header for FreeBSD

July 28, 2005
	acme: subtle winunlock bug fix
	libdiskfs: fix double, triple indirect block reading for FFS
	vftp: new program
	vbackup: coalesce mount-*BSD files.

July 27, 2005
	include/u.h, src/libthread/thread.c: SunOS 10 attempt
	libdraw: less graceful, less raceful shutdown
	libthread: avoid C library races during threadexitsall
	awd: cannot use hostname -s on SunOS
	9pserve, srv: use correct type for NOFID

July 26, 2005
	various: fixes for SunOS from Bengt Kleberg

July 25, 2005
	vac: avoid infinite redirection while sniffing type of archive

July 24, 2005
	venti: many minor changes, updates

src/libdisk/scsi.c
/*
 * Now thread-safe.
 *
 * The codeqlock guarantees that once codes != nil, that pointer will never 
 * change nor become invalid.
 *
 * The QLock in the Scsi structure moderates access to the raw device.
 * We should probably export some of the already-locked routines, but
 * there hasn't been a need.
 */

#include <u.h>
#include <libc.h>
#include <disk.h>

int scsiverbose;

#define codefile "/sys/lib/scsicodes"

static char *codes;
static QLock codeqlock;

static void
getcodes(void)
{
	Dir *d;
	int n, fd;

	if(codes != nil)
		return;

	qlock(&codeqlock);
	if(codes != nil) {
		qunlock(&codeqlock);
		return;
	}

	if((d = dirstat(codefile)) == nil || (fd = open(codefile, OREAD)) < 0) {
		qunlock(&codeqlock);
		return;
	}

	codes = malloc(1+d->length+1);
	if(codes == nil) {
		close(fd);
		qunlock(&codeqlock);
		free(d);
		return;
	}

	codes[0] = '\n';	/* for searches */
	n = readn(fd, codes+1, d->length);
	close(fd);
	free(d);

	if(n < 0) {
		free(codes);
		codes = nil;
		qunlock(&codeqlock);
		return;
	}
	codes[n] = '\0';
	qunlock(&codeqlock);
}
	
char*
scsierror(int asc, int ascq)
{
	char *p, *q;
	static char search[32];
	static char buf[128];

	getcodes();

	if(codes) {
		sprint(search, "\n%.2ux%.2ux ", asc, ascq);
		if(p = strstr(codes, search)) {
			p += 6;
			if((q = strchr(p, '\n')) == nil)
				q = p+strlen(p);
			snprint(buf, sizeof buf, "%.*s", (int)(q-p), p);
			return buf;
		}

		sprint(search, "\n%.2ux00", asc);
		if(p = strstr(codes, search)) {
			p += 6;
			if((q = strchr(p, '\n')) == nil)
				q = p+strlen(p);
			snprint(buf, sizeof buf, "(ascq #%.2ux) %.*s", ascq, (int)(q-p), p);
			return buf;
		}
	}

	sprint(buf, "scsi #%.2ux %.2ux", asc, ascq);
	return buf;
}


static int
_scsicmd(Scsi *s, uchar *cmd, int ccount, void *data, int dcount, int io, int dolock)
{
	uchar resp[16];
	int n;
	long status;

	if(dolock)
		qlock(&s->lk);
	if(write(s->rawfd, cmd, ccount) != ccount) {
		werrstr("cmd write: %r");
		if(dolock)
			qunlock(&s->lk);
		return -1;
	}

	switch(io){
	case Sread:
		n = read(s->rawfd, data, dcount);
		if(n < 0 && scsiverbose)
			fprint(2, "dat read: %r: cmd 0x%2.2uX\n", cmd[0]);
		break;
	case Swrite:
		n = write(s->rawfd, data, dcount);
		if(n != dcount && scsiverbose)
			fprint(2, "dat write: %r: cmd 0x%2.2uX\n", cmd[0]);
		break;
	default:
	case Snone:
		n = write(s->rawfd, resp, 0);
		if(n != 0 && scsiverbose)
			fprint(2, "none write: %r: cmd 0x%2.2uX\n", cmd[0]);
		break;
	}

	memset(resp, 0, sizeof(resp));
	if(read(s->rawfd, resp, sizeof(resp)) < 0) {
		werrstr("resp read: %r\n");
		if(dolock)
			qunlock(&s->lk);
		return -1;
	}
	if(dolock)
		qunlock(&s->lk);

	resp[sizeof(resp)-1] = '\0';
	status = atoi((char*)resp);
	if(status == 0)
		return n;

	werrstr("cmd %2.2uX: status %luX dcount %d n %d", cmd[0], status, dcount, n);
	return -1;
}

int
scsicmd(Scsi *s, uchar *cmd, int ccount, void *data, int dcount, int io)
{
	return _scsicmd(s, cmd, ccount, data, dcount, io, 1);
}

static int
_scsiready(Scsi *s, int dolock)
{
	uchar cmd[6], resp[16];
	int status, i;

	if(dolock)
		qlock(&s->lk);
	for(i=0; i<3; i++) {
		memset(cmd, 0, sizeof(cmd));
		cmd[0] = 0x00;	/* unit ready */
		if(write(s->rawfd, cmd, sizeof(cmd)) != sizeof(cmd)) {
			if(scsiverbose)
				fprint(2, "ur cmd write: %r\n");
			goto bad;
		}
		write(s->rawfd, resp, 0);
		if(read(s->rawfd, resp, sizeof(resp)) < 0) {
			if(scsiverbose)
				fprint(2, "ur resp read: %r\n");
			goto bad;
		}
		resp[sizeof(resp)-1] = '\0';
		status = atoi((char*)resp);
		if(status == 0 || status == 0x02) {
			if(dolock)
				qunlock(&s->lk);
			return 0;
		}
		if(scsiverbose)
			fprint(2, "target: bad status: %x\n", status);
	bad:;
	}
	if(dolock)
		qunlock(&s->lk);
	return -1;
}

int
scsiready(Scsi *s)
{
	return _scsiready(s, 1);
}

int
scsi(Scsi *s, uchar *cmd, int ccount, void *v, int dcount, int io)
{
	uchar req[6], sense[255], *data;
	int tries, code, key, n;
	char *p;

	data = v;
	SET(key); SET(code);
	qlock(&s->lk);
	for(tries=0; tries<2; tries++) {
		n = _scsicmd(s, cmd, ccount, data, dcount, io, 0);
		if(n >= 0) {
			qunlock(&s->lk);
			return n;
		}

		/*
		 * request sense
		 */
		memset(req, 0, sizeof(req));
		req[0] = 0x03;
		req[4] = sizeof(sense);
		memset(sense, 0xFF, sizeof(sense));
		if((n=_scsicmd(s, req, sizeof(req), sense, sizeof(sense), Sread, 0)) < 14)
			if(scsiverbose)
				fprint(2, "reqsense scsicmd %d: %r\n", n);
	
		if(_scsiready(s, 0) < 0)
			if(scsiverbose)
				fprint(2, "unit not ready\n");
	
		key = sense[2];
		code = sense[12];
		if(code == 0x17 || code == 0x18) {	/* recovered errors */
			qunlock(&s->lk);
			return dcount;
		}
		if(code == 0x28 && cmd[0] == 0x43) {	/* get info and media changed */
			s->nchange++;
			s->changetime = time(0);
			continue;
		}
	}

	/* drive not ready, or medium not present */
	if(cmd[0] == 0x43 && key == 2 && (code == 0x3a || code == 0x04)) {
		s->changetime = 0;
		qunlock(&s->lk);
		return -1;
	}
	qunlock(&s->lk);

	if(cmd[0] == 0x43 && key == 5 && code == 0x24)	/* blank media */
		return -1;

	p = scsierror(code, sense[13]);

	werrstr("cmd #%.2ux: %s", cmd[0], p);

	if(scsiverbose)
		fprint(2, "scsi cmd #%.2ux: %.2ux %.2ux %.2ux: %s\n", cmd[0], key, code, sense[13], p);

//	if(key == 0)
//		return dcount;
	return -1;
}

Scsi*
openscsi(char *dev)
{
	Scsi *s;
	int rawfd, ctlfd, l, n;
	char *name, *p, buf[512];

	l = strlen(dev)+1+3+1;
	name = malloc(l);
	if(name == nil)
		return nil;

	snprint(name, l, "%s/raw", dev);
	if((rawfd = open(name, ORDWR)) < 0) {
		free(name);
		return nil;
	}

	snprint(name, l, "%s/ctl", dev);
	if((ctlfd = open(name, ORDWR)) < 0) {
		free(name);
	Error:
		close(rawfd);
		return nil;
	}
	free(name);

	n = readn(ctlfd, buf, sizeof buf);
	close(ctlfd);
	if(n <= 0)
		goto Error;

	if(strncmp(buf, "inquiry ", 8) != 0 || (p = strchr(buf, '\n')) == nil)
		goto Error;
	*p = '\0';

	if((p = strdup(buf+8)) == nil)
		goto Error;

	s = malloc(sizeof(*s));
	if(s == nil) {
	Error1:
		free(p);
		goto Error;
	}
	memset(s, 0, sizeof(*s));

	s->rawfd = rawfd;
	s->inquire = p;
	s->changetime = time(0);
	
	if(scsiready(s) < 0)
		goto Error1;

	return s;
}
